File: scratch/inspect_commerce.py
import os
import json
import re

commerce_dir = '/Users/aryanmaurya/exam portal/COMMERCE_LATEX'
project_dir = '/Users/aryanmaurya/exam portal'
results = []

def walk(directory):
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith('.tex'):
                results.append(os.path.join(root, file))

walk(commerce_dir)
print(f"Found {len(results)} tex files.")

parsed_files = []

for file_path in results:
    relative_path = os.path.relpath(file_path, project_dir)
    file_name = os.path.basename(file_path)
    
    # Read the first 45 lines to search for titles/codes
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            lines = [f.readline() for _ in range(45)]
    except Exception as e:
        print(f"Error reading {file_name}: {e}")
        continue
        
    paper_title = ''
    paper_code = ''
    
    # Check lines for paper information
    for line in lines:
        if '% Paper:' in line:
            # e.g. % Paper: BCH-122 : Financial Accounting - II
            match = re.search(r'%\s*Paper:\s*([A-Za-z0-9\-_]+)\s*[:\-]\s*(.*)', line, re.IGNORECASE)
            if match:
                paper_code = match.group(1).strip()
                paper_title = match.group(2).strip()
                break
        elif 'Paper No.' in line:
            # e.g. {\large\bfseries Paper No. BCH-122: Financial Accounting - II}
            match = re.search(r'Paper\s+No\.\s*([A-Za-z0-9\-_]+)\s*[:\-]\s*([^\}\\]*)', line, re.IGNORECASE)
            if match:
                paper_code = match.group(1).strip()
                paper_title = match.group(2).strip()
                break
                
    # Normalize code and subject
    if paper_code:
        paper_code = paper_code.replace('-', '').strip()
        
    # Attempt to parse from filename if not found in headers
    # e.g., BCH214_CostAccounting_SemIII_2016-17.tex
    fn_match = re.match(r'^([A-Z0-9]+)_(.*?)_(Sem[IVXLC\d]+)_(\d{4}-\d{2})\.tex$', file_name, re.IGNORECASE)
    code_from_fn = ''
    subject_from_fn = ''
    sem_from_fn = ''
    year_from_fn = ''
    
    if fn_match:
        code_from_fn = fn_match.group(1)
        subject_from_fn = fn_match.group(2)
        # convert CamelCase to space separated
        subject_from_fn = re.sub(r'([A-Z])', r' \1', subject_from_fn).strip()
        sem_from_fn = fn_match.group(3)
        year_from_fn = fn_match.group(4)
    else:
        # backup match
        fn_match2 = re.match(r'^([A-Z0-9]+)_(.*?)_([A-Za-z0-9\-]+)\.tex$', file_name, re.IGNORECASE)
        if fn_match2:
            code_from_fn = fn_match2.group(1)
            subject_from_fn = fn_match2.group(2)
            subject_from_fn = re.sub(r'([A-Z])', r' \1', subject_from_fn).strip()
            
    final_code = paper_code if paper_code else code_from_fn
    final_subject = paper_title if paper_title else subject_from_fn
    
    # Normalize final_code by adding a hyphen after 3 letters, e.g. BCH121 -> BCH-121
    if final_code and len(final_code) >= 6 and not '-' in final_code:
        final_code = final_code[:3] + '-' + final_code[3:]
        
    # Semester mapping
    semester = 1
    sem_lower = (sem_from_fn if sem_from_fn else file_name).lower()
    if 'sem_i' in sem_lower or 'semi_' in sem_lower or 'sem-i' in sem_lower or 'semi' in sem_lower:
        semester = 1
    if 'sem_ii' in sem_lower or 'semii_' in sem_lower or 'sem-ii' in sem_lower or 'semii' in sem_lower:
        semester = 2
    if 'sem_iii' in sem_lower or 'semiii_' in sem_lower or 'sem-iii' in sem_lower or 'semiii' in sem_lower:
        semester = 3
    if 'sem_iv' in sem_lower or 'semiv_' in sem_lower or 'sem-iv' in sem_lower or 'semiv' in sem_lower:
        semester = 4
    if 'sem_v' in sem_lower or 'semv_' in sem_lower or 'sem-v' in sem_lower or 'semv' in sem_lower:
        semester = 5
    if 'sem_vi' in sem_lower or 'semvi_' in sem_lower or 'sem-vi' in sem_lower or 'semvi' in sem_lower:
        semester = 6

    final_year = year_from_fn if year_from_fn else ''
    if not final_year:
        year_match = re.search(r'(\d{4}-\d{2})', file_name)
        if year_match:
            final_year = year_match.group(1)

    parsed_files.append({
        'fileName': file_name,
        'filePath': relative_path,
        'code': final_code,
        'subject': final_subject,
        'semester': semester,
        'year': final_year
    })

# Sort by code and year
parsed_files.sort(key=lambda x: (x['code'], x['year']))

with open('/Users/aryanmaurya/exam portal/scratch/commerce_parsed.json', 'w', encoding='utf-8') as f:
    json.dump(parsed_files, f, indent=2)

print(f"Written parsed files to scratch/commerce_parsed.json")

File: scratch/inspect_commerce.mjs
import fs from 'fs';
import path from 'path';

const commerceDir = '/Users/aryanmaurya/exam portal/COMMERCE_LATEX';
const results = [];

function walk(dir) {
    const list = fs.readdirSync(dir);
    list.forEach(file => {
        const fullPath = path.join(dir, file);
        if (fs.statSync(fullPath).isDirectory()) {
            walk(fullPath);
        } else if (file.endsWith('.tex')) {
            results.push(fullPath);
        }
    });
}

walk(commerceDir);

console.log(`Found ${results.length} tex files.`);

const parsedFiles = results.map(filePath => {
    const relativePath = path.relative('/Users/aryanmaurya/exam portal', filePath);
    const fileName = path.basename(filePath);
    
    // Read the first 45 lines to search for titles/codes
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n').slice(0, 45);
    
    // Attempt to parse metadata from comments or text
    let paperLine = lines.find(l => l.includes('% Paper:') || l.includes('Paper No.'));
    let paperTitle = '';
    let paperCode = '';
    
    if (paperLine) {
        // e.g. % Paper: BCH-122 : Financial Accounting - II
        // or {\large\bfseries Paper No. BCH-122: Financial Accounting - II}
        const match = paperLine.match(/(?:Paper(?:\s+No\.)?:?\s*)([A-Z0-9\-_]+)\s*[:\-]\s*(.*)/i);
        if (match) {
            paperCode = match[1].trim();
            paperTitle = match[2].replace(/[\}\\]/g, '').trim();
        }
    }
    
    // If not found in those formats, parse from filename
    // e.g., BCH214_CostAccounting_SemIII_2016-17.tex
    const fnMatch = fileName.match(/^([A-Z0-9]+)_(.*?)_(Sem[IVXLC\d]+)_(\d{4}-\d{2})\.tex$/i);
    let codeFromFn = '';
    let subjectFromFn = '';
    let semFromFn = '';
    let yearFromFn = '';
    
    if (fnMatch) {
        codeFromFn = fnMatch[1];
        subjectFromFn = fnMatch[2].replace(/([A-Z])/g, ' $1').trim();
        semFromFn = fnMatch[3];
        yearFromFn = fnMatch[4];
    } else {
        // backup match
        const fnMatch2 = fileName.match(/^([A-Z0-9]+)_(.*?)_([A-Za-z0-9\-]+)\.tex$/i);
        if (fnMatch2) {
            codeFromFn = fnMatch2[1];
            subjectFromFn = fnMatch2[2].replace(/([A-Z])/g, ' $1').trim();
        }
    }
    
    // Normalize semester
    let semester = 1;
    const semLower = (semFromFn || fileName).toLowerCase();
    if (semLower.includes('sem_i') || semLower.includes('semi_') || semLower.includes('sem-i') || semLower.includes('semi')) semester = 1;
    if (semLower.includes('sem_ii') || semLower.includes('semii_') || semLower.includes('sem-ii') || semLower.includes('semii')) semester = 2;
    if (semLower.includes('sem_iii') || semLower.includes('semiii_') || semLower.includes('sem-iii') || semLower.includes('semiii')) semester = 3;
    if (semLower.includes('sem_iv') || semLower.includes('semiv_') || semLower.includes('sem-iv') || semLower.includes('semiv')) semester = 4;
    if (semLower.includes('sem_v') || semLower.includes('semv_') || semLower.includes('sem-v') || semLower.includes('semv')) semester = 5;
    if (semLower.includes('sem_vi') || semLower.includes('semvi_') || semLower.includes('sem-vi') || semLower.includes('semvi')) semester = 6;
    
    return {
        fileName,
        filePath: relativePath,
        code: paperCode || codeFromFn,
        subject: paperTitle || subjectFromFn,
        semester,
        year: yearFromFn || (fileName.match(/\d{4}-\d{2}/) ? fileName.match(/\d{4}-\d{2}/)[0] : '')
    };
});

fs.writeFileSync('/Users/aryanmaurya/exam portal/scratch/commerce_parsed.json', JSON.stringify(parsedFiles, null, 2));
console.log('Written parsed files to scratch/commerce_parsed.json');
